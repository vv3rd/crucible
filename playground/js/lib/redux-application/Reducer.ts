import { TaskScheduler } from "./Task";
import { Msg, MsgWith, SomeMsg } from "./Message";
import { isPlainObject } from "../toolkit";
import { Fn } from "./Fn";

export interface Reducer<TState, TMsg = SomeMsg> {
    (
        Reducer_currentState: TState | undefined,
        Reducer_message: TMsg,
        Reducer_taskScheduler: TaskScheduler<TState>,
    ): TState;
}

type AnyReducer = Reducer<any, any>;

export function Reducer<TState>(reducer: Reducer<TState>) {
    return reducer;
}

export namespace Reducer {
    const InitAction = { type: "INIT-" + Math.random() };

    export const build = createSlice;

    export function initialize<TState>(reducer: Reducer<TState>) {
        return reducer(undefined, InitAction, () => {});
    }

    export const primitive = createPrimitiveReducerImpl;

    export const compose = composeReducersImpl as <M extends Dict<AnyReducer>>(
        reducersDict: M,
    ) => ReducerFromCombination<M>;

    type ReducerFromCombination<
        M extends Dict<AnyReducer>,
        var_State = { [P in keyof M]: InferState<M[P]> },
    > = keyof M extends never ? Reducer<{}> : Reducer<var_State>;

    export type InferMsg<R> = R extends Reducer<any> ? Msg : never;
    export type InferState<R> = R extends Reducer<infer S> ? S : never;
}

function composeReducersImpl(reducersObject: Record<string, AnyReducer>) {
    type TState = any;
    type TMsg = any;

    const reducers = Object.entries(reducersObject);

    if (reducers.length === 0) {
        const emptyState = {};
        return () => emptyState;
    }

    return function composedReducer(
        current: TState | undefined,
        action: TMsg,
        schedule: TaskScheduler<TState>,
    ): TState {
        let next: TState = current;
        for (let [key, reducer] of reducers) {
            const scheduleScoped = TaskScheduler.scoped(schedule, (s) => s[key]);
            const stateWas = current?.[key];
            const stateNow = reducer(stateWas, action, scheduleScoped);
            if (stateWas !== stateNow) {
                if (next === current) {
                    next = { ...current };
                }
                next[key] = stateNow;
            }
        }
        return next;
    };
}

interface Accessor<T> {
    (): T;
    (Accessor_state: T | Partial<T>): T;
    (Accessor_update: (state: T) => T | Partial<T>): T;
    do: TaskScheduler<T>;
}

type Updaters<T> = {
    [key: string]: (this: Accessor<T>, ...args: any[]) => void | T;
};

const buildSlice =
    <TState>() =>
    <U extends Updaters<TState>>(updaters: U) => ({
        withInitialState: (initialState: TState) => ({
            withPrefix: <TName extends string>(name: TName) =>
                createSlice(name, initialState, updaters),
        }),
    });

type todo = { description: string; completed: boolean };

const itemsSlice = buildSlice<{ todos: todo[] }>()({
    itemAdded(item: todo) {},
    itemRemoved(item: todo) {},
    itemsCleared() {},
})
    .withInitialState({
        todos: [],
    })
    .withPrefix("todos");

export function createSlice<T, N extends string, R extends Updaters<T>>(
    reducerName: N,
    initialState: T,
    updaters: R,
) {
    type Addressed<K extends string> = `${N}/${K}`;
    type TMsgs = {
        [K in keyof R & string]: Msg.TypedFactory<
            Fn.Like<R[K], { returns: MsgWith<Parameters<R[K]>, Addressed<K>> }>
        >;
    };
    type TMsg = ReturnType<TMsgs[keyof TMsgs]>;

    const keys = Object.keys(updaters);
    const messages = Object.fromEntries(
        keys.map((key) => [key, Msg.ofType(`${reducerName}/${key}`).withPayload<any[]>()]),
    ) as unknown as TMsgs;

    const isMatchingMsg = (msg: Msg): msg is TMsg =>
        Object.values(messages).some(({ match }) => match(msg));

    const reducer: Reducer<T> = (state = initialState, msg, exec) => {
        if (!isMatchingMsg(msg)) {
            return state;
        }
        const key = msg.type.slice(reducerName.length + 1);
        const update = updaters[key];
        if (!update) {
            return state;
        }
        const accessor = ((...args) => {
            if (!args.length) {
                return state;
            }
            let next = args[0];
            if (next instanceof Function) next = next(state);
            if (isPlainObject(state) && isPlainObject(next)) {
                state = { ...state, ...next };
            } else {
                state = next as T;
            }
            return state;
        }) as Accessor<T>;
        accessor.do = exec;
        const outState = update.apply(accessor, msg.payload);
        if (outState !== undefined) {
            state = outState;
        }
        return state;
    };

    // TODO: add matchers for messages group, add getInitialState
    return Object.assign(reducer, {
        message: messages,
        reducer,
        reducerName,
    });
}

function createPrimitiveReducerImpl<T, S extends string>(initialState: T, updType: S) {
    const updateMsg = Msg.ofType(updType).withPayload<T | ((current: T) => T)>();

    const primitiveReducer = (state = initialState, msg: Msg) => {
        if (updateMsg.match(msg)) {
            if (msg.payload instanceof Function) {
                return msg.payload(state);
            } else {
                return msg.payload;
            }
        }
        return state;
    };
    primitiveReducer.update = updateMsg;
    return primitiveReducer;
}
