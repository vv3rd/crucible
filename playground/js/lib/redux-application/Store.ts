import { Message, Store, ListenerCallback } from "./types";
import { Reducer } from "./Reducer";
import { TaskTools, Task } from "./Task";
import { identity } from "../toolkit";
import { FUCK_INTERNALS_USED, FUCK_STORE_LOCKED } from "./Errors.ts";

type StoreOverlay = (creator: InnerStoreCreator) => InnerStoreCreator;

type InnerStoreCreator = <TState, TMsg extends Message>(
	reducer: Reducer<TState, TMsg>,
	internals: Internals<TState, TMsg>,
) => Store<TState, TMsg>;

type Internals<TState, TMsg extends Message> = {
	tasksExecutor: Executor;
	storeAccessor: () => Store<TState, TMsg>;
};

type Executor = <A, Fs extends Array<(arg: A) => any>>(
	funcs: Fs,
	param: A | (() => A),
) => ReturnType<Fs[0]>;

export const createStore = <TState, TMsg extends Message>(
	reducer: Reducer<TState, TMsg>,
	overlay: StoreOverlay = identity,
): Store<TState, TMsg> => {
	let storeAccessor = (): typeof store => {
		try {
			return store;
		} catch (error) {
			throw new Error(FUCK_INTERNALS_USED, { cause: error });
		}
	};
	let tasksExecutor: Executor = () => {
		throw new Error(FUCK_INTERNALS_USED);
	};
	const store: Store<TState, TMsg> = overlay(createStoreImpl)(reducer, {
		storeAccessor: () => storeAccessor(),
		tasksExecutor: (f, p) => tasksExecutor(f, p),
	});
	storeAccessor = () => store;
	tasksExecutor = defaultExecutor;
	return store;
};

// declare const testTask: Task<any, any, { foo: "test" }>;

const createStoreImpl: InnerStoreCreator = (reducer, internals) => {
	type TState = Reducer.InferState<typeof reducer>;
	type TMsg = Reducer.InferMsg<typeof reducer>;
	const {
		//
		tasksExecutor: executeTasks,
		storeAccessor: getFinalStore,
	} = internals;
	const listeners: Set<ListenerCallback<TMsg>> = new Set();
	const getTaskTools = () => TaskTools.fromStore(getFinalStore());

	let state: TState = Reducer.initialize(reducer);

	const activeStore: Store<TState, TMsg> = {
		dispatch(msgOrTask: Message<any> | Task<TState, any>) {
			if (typeof msgOrTask === "function") {
				const task = msgOrTask;
				return executeTasks([task], getTaskTools);
			}
			const msg = msgOrTask as TMsg;
			const tpb = Task.pool<TState, void>();
			try {
				delegate = lockedStore;
				state = reducer(state, msg, tpb.getScheduler());
			} finally {
				delegate = activeStore;
				tpb.lockScheduler();
			}
			executeTasks([...listeners], msg);
			executeTasks(tpb.getTasks(), getTaskTools);
		},

		subscribe(listener) {
			listeners.add(listener);
			const unsubscribe = () => delegate.unsubscribe(listener);
			return unsubscribe;
		},

		unsubscribe(listener) {
			listeners.delete(listener);
		},

		getState() {
			return state;
		},
	};

	let delegate: Store<TState, TMsg> = activeStore;

	// biome-ignore format: better visual
	return {
		dispatch: (action: any) => delegate.dispatch(action),
		getState:            () => delegate.getState(),
		subscribe:   (callback) => delegate.subscribe(callback),
		unsubscribe: (callback) => delegate.unsubscribe(callback),
	};
};

// biome-ignore format: saves space
const lockedStore: Store<any, any> = {
	dispatch() { throw new Error(FUCK_STORE_LOCKED); },
	getState() { throw new Error(FUCK_STORE_LOCKED); },
	subscribe() { throw new Error(FUCK_STORE_LOCKED); },
	unsubscribe() { throw new Error(FUCK_STORE_LOCKED); },
};

const defaultExecutor: Executor = (funcs, param) => {
	const onError = console.error;
	if (param instanceof Function) {
		param = param();
	}
	if (0 in funcs) {
		const firstFunc = funcs[0]!;
		try {
			var firstOut = firstFunc(param);
		} catch (err) {
			var firstErr = err;
		}
	}
	const errors: unknown[] = [];
	for (const func of funcs) {
		try {
			func(param);
		} catch (err) {
			errors.push(err);
		}
	}
	for (const error of errors) {
		onError(error);
	}
	if (firstErr) {
		throw firstErr;
	} else {
		return firstOut;
	}
};
