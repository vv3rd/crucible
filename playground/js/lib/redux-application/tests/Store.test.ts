import { test } from "bun:test";
import { createStore } from "../Store";
import { Reducer } from "../Reducer";
import { expect } from "bun:test";

const reducer = Reducer.compose({
    count: Reducer.primitive(0, "setCount"),
});

test("just stuff in general", () => {
    const store = createStore(reducer);
    expect(store.getState()).toMatchSnapshot();
    store.dispatch({ type: "setCount", payload: 10 });
    expect(store.getState()).toMatchSnapshot();
});

test("overlay", () => {
    const store = createStore(reducer, (createStore) => {
        return (reducer, final) => {
            const store = createStore(reducer, final);

            return {
                ...store,
            };
        };
    });
    //
    //
});
