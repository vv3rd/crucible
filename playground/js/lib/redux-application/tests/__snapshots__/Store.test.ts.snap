// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`just stuff in general 1`] = `
{
  "count": 0,
}
`;

exports[`just stuff in general 2`] = `
{
  "count": 10,
}
`;
